reading_comprehension/qanet_for_drop.py

```diff
@@ -0,0 +1,330 @@
+        # Shape: (batch_size,)
+        best_combination_prob = (best_signs_probs + (1 - numbers_mask) + 1e-10).log().sum(1).exp()
+        best_combination_prob *= answer_type_probs[:, 1]
+
+        # Shape: (batch_size,)
+        best_count_number = torch.argmax(count_number_probs, -1)
+        best_count_prob = torch.gather(count_number_probs, 1, best_count_number.unsqueeze(-1)).squeeze(-1)
+        best_count_prob *= answer_type_probs[:, 2]
+
+        best_answer_type = \
+            torch.argmax(torch.stack([best_span_prob, best_combination_prob, best_count_prob], 1), 1)
+
+        output_dict = {
+                "passage_question_attention": passage_question_attention,
+                "answer_type_probs": answer_type_probs,
+                "span_start_probs": span_start_probs,
+                "span_end_probs": span_end_probs,
+                "number_sign_probs": number_sign_probs,
+                "count_number_probs": count_number_probs,
+        }
+
+        # Compute the loss for training.
+        if answer_as_spans is not None:
+            # Shape: (batch_size, # of answer spans)
+            span_starts = answer_as_spans[:, :, 0]
+            span_ends = answer_as_spans[:, :, 1]
+            # Some spans are padded with index -1, so we need to mask them
+            span_mask = (span_starts != -1).float()
+            clamped_span_starts = torch.nn.functional.relu(span_starts)
+            clamped_span_ends = torch.nn.functional.relu(span_ends)
+            # Shape: (batch_size, # of answer spans)
+            likelyhood_for_span_starts = torch.gather(span_start_probs, 1, clamped_span_starts)
+            likelyhood_for_span_ends = torch.gather(span_end_probs, 1, clamped_span_ends)
+            # Shape: (batch_size, # of answer spans)
+            likelyhood_for_spans = likelyhood_for_span_starts * likelyhood_for_span_ends
+            likelyhood_for_spans = likelyhood_for_spans * span_mask
+
+            # Some combinations are padded with the labels of all numbers as 0, so we mask them here
+            # Shape: (batch_size, # of combinations)
+            combination_mask = (answer_as_plus_minus_combinations.sum(-1) > 0).float()
+            # Shape: (batch_size, # of numbers in the passage, # of combinations)
+            combinations = answer_as_plus_minus_combinations.transpose(1, 2)
+            # Shape: (batch_size, # of numbers in the passage, # of combinations)
+            likelyhood_for_combination_numbers = torch.gather(number_sign_probs, 2, combinations)
+            # the likelyhood of the masked positions should be 1 so that it will not affect the joint probability
+            likelyhood_for_combination_numbers = \
+                likelyhood_for_combination_numbers * numbers_mask.unsqueeze(-1) + (1 - numbers_mask.unsqueeze(-1))
+
+            # Shape: (batch_size, # of combinations)
+            likelyhood_for_combinations = \
```

Comment:
This is not numerically stable.  If you really want to work in probability space, use `util.logsumexp()`.  But really, all of this logic should be done in log space, so you don't have underflow problems.

Reply:
I'm not sure about this suggestion. `util.logsumexp()` equals to `tensor.exp().sum().log()`, while I want to do `tensor.log().sum().exp()`. This is because I need to multiply those likelihoods, and it's easier for pytorch to do it if I transform multiply to addition using `log`. I cannot keep those likelihoods in log space, since I need to add those likelihoods when computing the marginal likelihood (lines 254~257). Does this make sense? Any further suggestions?

Reply:
Oh, sorry, you're right about this line.  But you should definitely just keep everything in log space.

You should use `util.logsumexp()` when you need to _add_ log numbers in probability space.  So, lines 254-257 should use `util.logsumexp()` to do the sum.

Reply:
Basically, when you go from probability space to log space, you replace `*` everywhere with `+`, and `+` with `logsumexp()`.

Reply:
But what I need to do in lines 254-257 is to add probabilities rather than log probabilities, right? The loss is a negtive marginal likelihood, like:
```
loss = - log( P(type=span) * \sum P(span_i) + P(type=number) * \sum P(signs_for_numbers) + P(type=count) * \sum P(count) )
```

Reply:
Yes, and the way that you add probabilities, when the probabilities are given in log space, is to use `logsumexp()`:

```python
# all of the likelihood_for_* variables contain _log probabilities_
marginal_span_likelihood = answer_type_probs[:, 0] + util.logsumexp(likelihood_for_spans, dim=-1)
marginal_combination_likelihood = answer_type_probs[: 1] + util.logsumexp(likelihood_for_combinations, dim=-1)
marginal_count_likelihood = answer_type_probs[:, 2] + util.logsumexp(likelihood_for_counts, dim=-1)
marginal_likelihood = util.logsumexp(torch.stack([marginal_span_likelihood, marginal_combination_likelihood, marginal_count_likelihood], dim=-1), dim=-1)
```

At least that's the basic idea; might be some bugs in there.

Reply:
Oh, I see the point here. Thx!